### PR TITLE
Feature/project management

### PR DESCRIPTION
**Pull request by yasserahmed10**
This pull request introduces a comprehensive project management system with repository implementations, use cases, and corresponding tests. The changes include adding CRUD operations for projects, implementing various use cases for project management, defining custom exceptions, and ensuring robust test coverage.

### Repository and Model Enhancements:
* Implemented `ProjectsRepositoryImpl` to manage projects in memory with CRUD operations (`add`, `update`, `delete`, `getAll`, `getById`) (`src/main/kotlin/data/repositories/ProjectsRepositoryImpl.kt`).
* Updated `ProjectsRepository` interface to define the contract for project-related operations (`src/main/kotlin/logic/repositoies/ProjectsRepository.kt`).

### Use Case Implementations:
* Added `CreateProjectUseCase`, `DeleteProjectUseCase`, `GetAllProjectsUseCase`, `GetProjectByIdUseCase`, and `UpdateProjectUseCase` to encapsulate business logic for project management (`src/main/kotlin/logic/usecase`) [[1]](diffhunk://#diff-6990946da49f22d2a0a4106ac61b298fabad32370fb2e63f543ff2b29713aa26R1-R40) [[2]](diffhunk://#diff-e52073f1c203524e7692ae1bfe0fda60edb0754ab08ada4791584417104ef2f3R1-R19) [[3]](diffhunk://#diff-10dd5903351a537e24604f4aee27f124e6d063a32e8bb732048e90c4322bdf56R1-R13) [[4]](diffhunk://#diff-2ded9cf1aa1c19d08800957c85088be5cd3421cb447b76c98f84ecbd3156d229R1-R14) [[5]](diffhunk://#diff-fdcefac54d91e142e56362e030eb6e783bccc598ab6f644b6817f033695203bcR1-R36).

### Exception Handling:
* Defined custom exceptions (`NotAdminException`, `InvalidProjectNameException`, `ProjectCreationFailedException`, `ProjectNotFoundException`, `NoExistProjectException`) to handle specific error cases in project operations (`src/main/kotlin/utilities/Exception.kt`).

### Test Coverage:
* Added unit tests for all use cases to ensure correctness and handle edge cases:
  - `CreateProjectUseCaseTest` verifies project creation scenarios, including validation and repository interactions (`src/test/kotlin/logic/usecase/CreateProjectUseCaseTest.kt`).
  - `DeleteProjectUseCaseTest` tests project deletion logic, including admin checks and non-existent projects (`src/test/kotlin/logic/usecase/DeleteProjectUseCaseTest.kt`).
  - `GetAllProjectsUseCaseTest` validates retrieval of all projects from the repository (`src/test/kotlin/logic/usecase/GetAllProjectsUseCaseTest.kt`).
  - `GetProjectByIdUseCaseTest` ensures correct project retrieval by ID and handles missing projects (`src/test/kotlin/logic/usecase/GetProjectByIdUseCaseTest.kt`).
  - `UpdateProjectUseCaseTest` validates project updates, including name validation and admin checks (`src/test/kotlin/logic/usecase/UpdateProjectUseCaseTest.kt`).

### CODE REVIEW

**Review by ZeyadAbdullah679 — CHANGES_REQUESTED** [review]
You have to create test cases for `ProjectsRepositoryImpl` class

**Comment by yasserahmed10** [comment]
@ZeyadAbdullah679 
working on it

**Comment by yasserahmed10** [comment]
@ZeyadAbdullah679 
Done

File: src/main/kotlin/logic/usecase/GetAllProjectsUseCase.kt [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package logic.usecase
+
+import logic.models.Project
+import logic.repositoies.ProjectsRepository
+
+class GetAllProjectsUseCase(
+    private val projectsRepository: ProjectsRepository
+) {
+
+    fun invoke(): List<Project> {
```
Comment (ZeyadAbdullah679) [inline src/main/kotlin/logic/usecase/GetAllProjectsUseCase.kt]: Put `operator` before `fun invoke()` for all use cases and replace the calling to the use case by this `GetAllProjectUseCase(projectRepository)` not only this use case for all use cases do that

File: src/main/kotlin/logic/usecase/DeleteProjectUseCase.kt [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import logic.repositoies.ProjectsRepository
+import utilities.NoExistProjectException
+import utilities.NotAdminException
+import java.util.UUID
+
+class DeleteProjectUseCase(private val projectsRepository: ProjectsRepository) {
+    fun invoke(id: UUID, isAdmin: Boolean): Boolean {
```
Comment (ZeyadAbdullah679) [inline src/main/kotlin/logic/usecase/DeleteProjectUseCase.kt]: Put `operator` before `fun invoke()` for all use cases and replace the calling to the use case by this `GetAllProjectUseCase(projectRepository)` not only this use case for all use cases do that

File: src/main/kotlin/logic/usecase/CreateProjectUseCase.kt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import logic.repositoies.ProjectsRepository
+import utilities.InvalidProjectNameException
+import utilities.NotAdminException
+import utilities.ProjectCreationFailedException
+import java.util.UUID
+
+class CreateProjectUseCase(
+    private val projectsRepository: ProjectsRepository
+) {
+    fun invoke(name: String, userIds: List<UUID>, isAdmin: Boolean): UUID {
```
Comment (ZeyadAbdullah679) [inline src/main/kotlin/logic/usecase/CreateProjectUseCase.kt]: Put `operator` before `fun invoke()` for all use cases and replace the calling to the use case by this `GetAllProjectUseCase(projectRepository)` not only this use case for all use cases do that

File: src/main/kotlin/logic/usecase/GetProjectByIdUseCase.kt [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+import logic.models.Project
+import logic.repositoies.ProjectsRepository
+import java.util.UUID
+
+class GetProjectByIdUseCase(
+    private val projectsRepository: ProjectsRepository
+) {
+
+    fun invoke(id: UUID): Project {
```
Comment (ZeyadAbdullah679) [inline src/main/kotlin/logic/usecase/GetProjectByIdUseCase.kt]: Put `operator` before `fun invoke()` for all use cases and replace the calling to the use case by this `GetAllProjectUseCase(projectRepository)` not only this use case for all use cases do that

File: src/main/kotlin/logic/usecase/UpdateProjectUseCase.kt [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import utilities.InvalidProjectNameException
+import utilities.NoExistProjectException
+import utilities.NotAdminException
+import java.util.UUID
+
+class UpdateProjectUseCase(
+    private val projectsRepository: ProjectsRepository
+) {
+
+    fun invoke(id: UUID, name: String, userIds: List<UUID>, isAdmin: Boolean): Boolean {
```
Comment (ZeyadAbdullah679) [inline src/main/kotlin/logic/usecase/UpdateProjectUseCase.kt]: Put `operator` before `fun invoke()` for all use cases and replace the calling to the use case by this `GetAllProjectUseCase(projectRepository)` not only this use case for all use cases do that

**Review by ZeyadAbdullah679 — CHANGES_REQUESTED** [review]
@yasserahmed10

File: src/main/kotlin/logic/usecase/GetAllProjectsUseCase.kt (hunk 1 above)
Comment (yasserahmed10) [reply src/main/kotlin/logic/usecase/GetAllProjectsUseCase.kt]: Done

File: src/main/kotlin/logic/usecase/DeleteProjectUseCase.kt (hunk 2 above)
Comment (yasserahmed10) [reply src/main/kotlin/logic/usecase/DeleteProjectUseCase.kt]: Done

**Review by baraa0abd — APPROVED** [review]
well done 🎉

File: src/main/kotlin/logic/usecase/CreateProjectUseCase.kt (hunk 3 above)
Comment (yasserahmed10) [reply src/main/kotlin/logic/usecase/CreateProjectUseCase.kt]: Done

File: src/main/kotlin/logic/usecase/GetProjectByIdUseCase.kt (hunk 4 above)
Comment (yasserahmed10) [reply src/main/kotlin/logic/usecase/GetProjectByIdUseCase.kt]: Done

File: src/main/kotlin/logic/usecase/UpdateProjectUseCase.kt (hunk 5 above)
Comment (yasserahmed10) [reply src/main/kotlin/logic/usecase/UpdateProjectUseCase.kt]: Done

**Comment by yasserahmed10** [comment]
@ZeyadAbdullah679 
you can merge now

**Review by ZeyadAbdullah679 — APPROVED** [review]
الله ينور يا international
